### PR TITLE
ASoC: Intel: consistent HDMI codec probing code

### PR DESCRIPTION
**Pull request by lyakh**
Consistent list handling and clean up of multiple .late_probe() machine driver methods.

### CODE REVIEW

**Comment by plbossart** [comment]
I'll defer comments to @kv2019i 
I wonder if we shouldn't use a helper instead of doing the same things in multiple drivers?

**Comment by kv2019i** [comment]
@plbossart wrote:
> I'll defer comments to @kv2019i
> I wonder if we shouldn't use a helper instead of doing the same things in multiple drivers?

Thanks @lyakh for doing the cleanups. I kept the mach drv changes to a minimum in my original common hdmi patchset as those were in review&testing for a long time and needed frequent rebases. I did put the new code to helper functions in hda_dsp_common.[ch] files for the common parts, but left old hdac-based code untouched for most part.

The jury is still not fully out on how long we need to maintain hdac-hdmi support. One easy option is to just remove the old hdac setup code. OTOH, that probably won't happen immediately (affects also SST) so a cleanup of the old code still makes sense. The only bit that is not exactly the same is how the private pcm handle in managed (e.g. "struct tgl_hdmi_pcm") and whether the machine driver support both HDMI (new ones will only support common hdmi). There's some variation related to this across the machine drivers, but the code is still pretty much the same, so helpers can be written.

File: sound/soc/intel/boards/bxt_da7219_max98357a.c [hunk 1]
```diff
@@ -617,12 +617,15 @@ static int bxt_card_late_probe(struct snd_soc_card *card)
 		snd_soc_dapm_add_routes(&card->dapm, broxton_map,
 					ARRAY_SIZE(broxton_map));
 
-	pcm = list_first_entry(&ctx->hdmi_pcm_list, struct bxt_hdmi_pcm,
-			       head);
-	component = pcm->codec_dai->component;
+	if (list_empty(&ctx->hdmi_pcm_list))
+		return -EINVAL;
```
Comment (kv2019i) [inline sound/soc/intel/boards/bxt_da7219_max98357a.c:621]: This is basicly hardcoded to the machine drivers DAI list, so in practise can't happen. But hmm, especially if this is pulled to a common function, better to have the check still.

File: sound/soc/intel/boards/bxt_da7219_max98357a.c [hunk 2]
```diff
@@ -617,12 +617,15 @@ static int bxt_card_late_probe(struct snd_soc_card *card)
...
 
-	if (ctx->common_hdmi_codec_drv)
+	if (ctx->common_hdmi_codec_drv) {
+		pcm = list_first_entry(&ctx->hdmi_pcm_list, struct bxt_hdmi_pcm,
+				       head);
+		component = pcm->codec_dai->component;
 		return hda_dsp_hdmi_build_controls(card, component);
+	}
```
Comment (kv2019i) [inline sound/soc/intel/boards/bxt_da7219_max98357a.c:628]: Maybe we would have a common macro to dig up the component? Only thing that varies is the struct field name (here struct bxt_hdmi_pcm).

File: sound/soc/intel/boards/bxt_da7219_max98357a.c [hunk 3]
```diff
@@ -643,9 +646,6 @@ static int bxt_card_late_probe(struct snd_soc_card *card)
 		i++;
 	}
```
Comment (kv2019i) [inline sound/soc/intel/boards/bxt_da7219_max98357a.c:648]: And the above hdac-specific setting could be put to hda_dsp_common.h. I didn't put it in my original series as I wanted to keep the series as compact as possible, but would reduce overlapping code. Can be done in a separate PR as well.

File: sound/soc/intel/boards/cml_rt1011_rt5682.c [hunk 4]
```diff
@@ -241,12 +241,15 @@ static int sof_card_late_probe(struct snd_soc_card *card)
...
+	if (list_empty(&ctx->hdmi_pcm_list))
+		return -EINVAL;
 
-	if (ctx->common_hdmi_codec_drv)
+	if (ctx->common_hdmi_codec_drv) {
+		pcm = list_first_entry(&ctx->hdmi_pcm_list, struct hdmi_pcm,
+				       head);
+		component = pcm->codec_dai->component;
 		return hda_dsp_hdmi_build_controls(card, component);
+	}
```
Comment (kv2019i) [inline sound/soc/intel/boards/cml_rt1011_rt5682.c:252]: Ok with the change could be put to a helper.

**Comment by plbossart** [comment]
let's add a helper later.

**Pull request merged**